src/collections.h

```diff
@@ -119,6 +128,34 @@ Collection<T>::fold(std::function<T(T, T)> func) {
 
 
 template<typename T>
+template<typename U>
+Collection<std::pair<T, U>>
+Collection<T>::zip(Collection<U> other_list) {
+    // TODO: list size checking 
+    using return_type = std::pair<T, U>;
+
+    std::vector<return_type> list(Data.size());
+    for (int i = 0; i < Data.size(); i++)
+        list[i] = std::make_pair(Data[i], other_list[i]);
+    return Collection<return_type>(list);
+};
+
+
+template<typename T>
+template<typename Function, typename U>
+Collection<typename std::result_of<Function(T, U)>::type>
```

Comment:
My understanding of zipWith (which comes from [here](http://hackage.haskell.org/package/base-4.8.0.0/docs/Prelude.html#v:zipWith)) is that it returns a single list, which is the result of the function called pairwise on the two lists.

for example:
```
list a = [1,2,3]
list b = [1,2,3]
list c = a.zipWith(sum, b) = [1+1, 2+2, 3+3]
```